File: code/pipeline-rust/src/thread/inference.rs
use std::{
    hint::spin_loop,
    sync::{Arc, Mutex, mpsc::SyncSender},
    thread::{self, JoinHandle},
};

use anyhow::{Context, Result};

use crate::{
    inference::InferenceEngine,
    os::{ShmBuffer, ShmFrame, now_nanos},
    queue::Queue,
};

/// Spawns a new thread that continuously processes frames from a shared memory
/// buffer and sends them to the next stage in the pipeline, running inference
/// using the specified ONNX model.
///
/// * `stream_name` - The name of the stream associated with this inference
///   thread.
/// * `queue` - A reference to the Queue from which frames will be popped for
///   processing.
/// * `sender` - A reference to the Sender used to send processed frames to the
///   next stage in the pipeline.
/// * `model_path` - The path to the ONNX model file used for inference.
/// * `window_frames` - The number of frames to process in each inference
///   window.
/// * `frame_shape` - The shape of the frames being processed, suitable for
///   tensor.
/// * `item_size_bytes` - The size of each frame item in bytes (e.g., 1 byte for
///   u8, 4 bytes for f32).
///
/// Returns a `Result` containing the `JoinHandle` of the spawned thread, or an
/// error if the thread could not be spawned.
pub fn spawn_inference_thread(
    stream_name: impl Into<String>,
    queue: &Arc<Mutex<Queue<ShmFrame>>>,
    sender: SyncSender<ShmFrame>,
    model_path: &str,
    window_frames: usize,
    frame_shape: Vec<i64>,
    item_size_bytes: usize,
) -> Result<JoinHandle<()>> {
    let stream_name = stream_name.into();
    let queue = Arc::clone(queue);
    let model_path = model_path.to_string();

    thread::Builder::new()
        .name(format!("inference_{}", stream_name))
        .spawn(move || {
            let window_size_items: usize =
                frame_shape.iter().map(|&s| s as usize).product();
            let frame_size_items = window_size_items / window_frames;

            let mut engine = InferenceEngine::try_new(&model_path, frame_shape)
                .expect("Failed to initialise InferenceEngine");

            let mut samples_collected = 0;

            loop {
                let (item, lapped_frames, dropped_frames) = {
                    let mut q = queue.lock().unwrap();
                    (q.pop(), q.lapped_frames, q.dropped_frames)
                };

                let t_pipeline_in = now_nanos().expect(
                    "Failed to get current time in \
                        nanoseconds for t_pipeline_in",
                );

                if let Some(mut frame) = item {
                    frame.lapped_frames = lapped_frames;
                    frame.dropped_frames = dropped_frames;

                    if frame.seq_num == u64::MAX {
                        // The generator stream has ended, so we send the final
                        // frame to the fusion thread and exit the loop.
                        sender.send(frame).expect(
                            "Failed to send exit signal frame to output queue",
                        );
                        break;
                    }

                    let item_offset = samples_collected * frame_size_items;

                    if item_size_bytes == 1 {
                        unsafe {
                            let src_slice = std::slice::from_raw_parts(
                                frame.payload_ptr.ptr,
                                frame_size_items,
                            );

                            let dest_slice = &mut engine.input_buffer_mut()
                                [item_offset..item_offset + frame_size_items];

                            for (d, s) in
                                dest_slice.iter_mut().zip(src_slice.iter())
                            {
                                *d = *s as f32;
                            }
                        }
                    } else {
                        unsafe {
                            let frame_size_bytes =
                                frame_size_items * std::mem::size_of::<f32>();

                            let dest_ptr = engine
                                .input_buffer_mut()
                                .as_mut_ptr()
                                .add(item_offset)
                                as *mut u8;

                            std::ptr::copy_nonoverlapping(
                                frame.payload_ptr.ptr,
                                dest_ptr,
                                frame_size_bytes,
                            );
                        }
                    }

                    samples_collected += 1;

                    if samples_collected >= window_frames {
                        frame.timestamps[ShmBuffer::PIPELINE_IN_TS] =
                            t_pipeline_in;

                        frame.inference_result =
                            engine.run().expect("Inference execution failed");

                        frame.timestamps[ShmBuffer::PIPELINE_OUT_TS] =
                            now_nanos().expect(
                                "Failed to get current time in \
                            nanoseconds for t_pipeline_out",
                            );

                        sender
                            .send(frame)
                            .expect("Failed to send frame to output queue");

                        samples_collected = 0;
                    }
                } else {
                    // Yield the thread to avoid busy waiting when the queue is
                    // empty.
                    spin_loop();
                }
            }
        })
        .with_context(|| {
            format!("Failed to spawn inference thread for '{stream_name}'")
        })
}

File: code/pipeline-cpp/src/thread/inference.cpp
#include <cstring>
#include <numeric>

#include <inference/inference_engine.h>
#include <os/os.h>
#include <os/shm.h>
#include <os/time.h>

#include "inference.h"

std::jthread
spawn_inference_thread(const std::string& stream_name,
  std::shared_ptr<Queue<ShmBuffer::Frame>> queue,
  Sender<ShmBuffer::Frame> sender,
  const std::string& model_path,
  const size_t window_frames,
  const std::vector<int64_t>& frame_shape,
  const size_t item_size_bytes)
{
  // clang-format off
  auto thread = std::jthread([stream_name, sender, queue, model_path,
    window_frames, frame_shape, item_size_bytes]() mutable {
    // clang-format on

    pthread_setname_np(pthread_self(),
      std::format("inference_{}", stream_name).substr(0, 15).c_str());

    const size_t window_size_items = static_cast<size_t>(std::accumulate(
      frame_shape.begin(), frame_shape.end(), 1LL, std::multiplies<int64_t>()));
    const size_t frame_size_items = window_size_items / window_frames;

    std::vector<float> tensor_data(window_size_items, 0.0f);
    InferenceEngine engine(model_path, frame_shape, tensor_data.data());

    size_t samples_collected = 0;

    for (;;) {
      std::unique_lock<std::mutex> transaction_lock(queue->get_mutex());
      auto item = queue->pop();
      const auto t_pipeline_in = current_time_nanos();
      auto lapped_frames = queue->get_lapped_frames();
      auto dropped_frames = queue->get_dropped_frames();
      transaction_lock.unlock();

      if (item.has_value()) {
        item->lapped_frames = lapped_frames;
        item->dropped_frames = dropped_frames;

        if (item->seq_num == ShmBuffer::Header::POISON_PILL) {
          // The generator stream has ended, so we send the final
          // frame to the fusion thread and exit the loop.
          sender.send(std::move(item.value()));
          break;
        }

        const size_t item_offset = samples_collected * frame_size_items;

        if (item_size_bytes == 1) {
          const uint8_t* raw_src =
            reinterpret_cast<const uint8_t*>(item->payload_ptr);
          for (size_t i = 0; i < frame_size_items; ++i) {
            tensor_data[item_offset + i] = static_cast<float>(raw_src[i]);
          }
        } else {
          const size_t frame_size_bytes = frame_size_items * sizeof(float);
          std::memcpy(tensor_data.data() + item_offset, 
            item->payload_ptr, 
            frame_size_bytes);
        }

        samples_collected++;

        if (samples_collected >= window_frames) {
          item->timestamps[ShmBuffer::PIPELINE_IN_TS] = t_pipeline_in;

          const auto& out = engine.run();
          std::copy(out.begin(), out.end(), std::begin(item->inference_result));

          item->timestamps[ShmBuffer::PIPELINE_OUT_TS] = current_time_nanos();

          try {
            sender.send(std::move(item.value()));
          } catch (const std::exception& e) {
            throw std::runtime_error(std::format(
              "Failed to send frame to output queue: {}", e.what()));
          }

          samples_collected = 0;
        }
      } else {
        spin_loop();
      }
    }
  });

  return thread;
}
